src/core/server/saved_objects/routes/export.ts

```diff
@@ -24,14 +24,109 @@ import { createPromiseFromStreams, createMapStream, createConcatStream } from '@
 
 interface RouteDependencies {
   config: SavedObjectConfig;
   coreUsageData: CoreUsageDataSetup;
 }
 
+type EitherExportOptions = SavedObjectsExportByTypeOptions | SavedObjectsExportByObjectOptions;
+
+interface ExportRawOptions {
+  type?: string | string[];
+  hasReference?: { id: string; type: string } | Array<{ id: string; type: string }>;
+  objects?: Array<{ id: string; type: string }>;
+  search?: string;
+  includeReferencesDeep: boolean;
+  excludeExportDetails: boolean;
+}
+
+interface ExportOptions {
+  types?: string[];
+  hasReference?: Array<{ id: string; type: string }>;
+  objects?: Array<{ id: string; type: string }>;
+  search?: string;
+  includeReferencesDeep: boolean;
+  excludeExportDetails: boolean;
+}
+
+const cleanOptions = ({
+  type,
+  objects,
+  search,
+  hasReference,
+  excludeExportDetails,
+  includeReferencesDeep,
+}: ExportRawOptions): ExportOptions => {
+  return {
+    types: typeof type === 'string' ? [type] : type,
+    search,
+    objects,
+    hasReference: hasReference && !Array.isArray(hasReference) ? [hasReference] : hasReference,
+    excludeExportDetails,
+    includeReferencesDeep,
+  };
+};
+
+const isExportByTypeOptions = (
+  options: EitherExportOptions
+): options is SavedObjectsExportByTypeOptions => {
+  return Boolean((options as SavedObjectsExportByTypeOptions).types);
+};
+
+const validateOptions = (
+  {
+    types,
+    objects,
+    excludeExportDetails,
+    hasReference,
+    includeReferencesDeep,
+    search,
+  }: ExportOptions,
+  { exportSizeLimit, supportedTypes }: { exportSizeLimit: number; supportedTypes: string[] }
+): EitherExportOptions => {
+  const hasTypes = (types?.length ?? 0) > 0;
+  const hasObjects = (objects?.length ?? 0) > 0;
+  if (hasTypes && hasObjects) {
+    throw new Error(`Can't specify both "types" and "objects" properties when exporting`);
+  } else if (hasObjects) {
+    if (objects!.length > exportSizeLimit) {
+      throw new Error(`Can't export more than ${exportSizeLimit} objects`);
+    }
+    if (typeof search === 'string') {
+      throw new Error(`Can't specify both "search" and "objects" properties when exporting`);
+    }
+    if (hasReference && hasReference.length) {
+      throw new Error(`Can't specify both "references" and "objects" properties when exporting`);
+    }
+    const validationError = validateObjects(objects!, supportedTypes);
+    if (validationError) {
+      throw new Error(validationError);
+    }
+    return {
+      objects: objects!,
+      excludeExportDetails,
+      includeReferencesDeep,
+    };
+  } else if (hasTypes) {
+    const validationError = validateTypes(types!, supportedTypes);
+    if (validationError) {
+      throw new Error(validationError);
+    }
+    return {
+      types: types!,
+      hasReference,
+      search,
+      excludeExportDetails,
+      includeReferencesDeep,
+    };
+  } else {
+    throw new Error('Either `type` or `objects` are required.');
```

Comment:
could you move it closer to if `(hasTypes && hasObjects) {` to handler edge cases earlier? as a bonus: we can get rid of nested `else if`